(Closed draft) Sites dashboard: Add badge and style tabs

sejas opened the pull request:
#### Proposed Changes

* Fix coming soon filtering
* Style Tabs
* Add new `Badge` component
* [ ] Add the components to Documentation
* [ ] Fix title types, from `string` to `Element`
* [ ] Improve performance adding `useMemo` or equivalent.


#### Screenshots
<img width="952" alt="Screenshot 2022-07-04 at 19 45 05" src="https://user-images.githubusercontent.com/779993/177200146-4e83e031-993a-4cda-b67c-7c69b8572f93.png">

https://user-images.githubusercontent.com/779993/177200898-77e63925-5bd9-4417-be52-29494e6068de.mp4



#### Testing Instructions

* Access to : /sites-dashboard (https://container-friendly-gauss.calypso.live/sites-dashboard)
* Observe the tabs match the style and functionality.

- Solves #65165

p-jackson committed:
The __i18n_text_domain__ constant so it can be used by shared components

When components in packages use the __() function, they don't know what
text domain to use for the second parameter. That's because they're a
shared package; they don't know what "domain" they're going to be used
in!

The way we get around this is to use the __i18n_text_domain__ build-time
constant as the text domain. This allows the package to compile without
errors, but when the app using the component gets built the
__i18n_text_domain__ constant gets replaced with the actual text domain
e.g. 'full-site-editing` or 'default'

p-jackson committed:
Add missing attributes to the the SiteData type definition

p-jackson committed:
Define the <SitesTable> component

This component displays a list of sites in a table. It is mostly
unstyled at the moment, but more is coming soon.

The component doesn't know how to do any filtering or fetching, which
means it should be reusable in multiple contexts.

p-jackson committed:
Define the <SitesTable> component

This component displays a list of sites in a table. It is mostly
unstyled at the moment, but more is coming soon.

The component doesn't know how to do any filtering or fetching, which
means it should be reusable in multiple contexts.

p-jackson committed:
Use Emotion for the global style overrides

p-jackson committed:
Define the <SitesTableFilterTabs> component

The tab component is separate from the main <SitesTable> component so
that places that don't need filtering won't have to include it.

p-jackson committed:
Define the <SitesTableFilterTabs> component

The tab component is separate from the main <SitesTable> component so
that places that don't need filtering won't have to include it.

p-jackson committed:
Display table, tabs, and button in the sites dashboard

sejas committed:
Sites Dashboard: add badge component for tabs

sejas committed:
Sites Dashboard: add badge component for tabs

sejas committed:
Sites Dashboard: fix coming soon filter

sejas committed:
Sites Dashboard: fix coming soon filter

sejas committed:
Sites Dashboard: style tab button and use Badge component

sejas committed:
Sites Dashboard: style tab button and use Badge component

sejas commented:
I need to update the typescript types or add comments to ignore the TS validation.
https://github.dev/WordPress/gutenberg/blob/trunk/packages/components/src/tab-panel/index.js#L21-L33

The `TabButton` uses a children to display the content inside itself.

But the types doesn't support a React Element, the title is defined only as string.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/tab-panel/index.d.ts#L12

p-jackson commented:
We're probably not going to be able to update the type in the DefinitelyTyped project because Gutenberg documents that type as a string. https://github.com/WordPress/gutenberg/blob/d923f2cae92f809541b7e009d68fd49ea3ea40ea/packages/components/src/tab-panel/README.md?plain=1#L121
But who knows, maybe we could submit a PR to change that documentation.

When I saw that the `<TabPanel>` component didn't support arbitrary elements for the title it made me question whether we'd be able to use this component 🤔 

But in the meantime, this is how I've hacked my way around "incorrect" type definitions in the past. Ideally we'd fix the issue upstream and then be able to get rid of this hack.

```typescript
interface MyTab extends Omit< TabPanel.Tab, 'title' > {
	title: React.ReactChild;
}

interface MyTabPanelProps extends Omit< TabPanel.Props, 'children' | 'tabs' > {
	tabs: readonly MyTab[];
	children: ( tab: MyTab ) => JSX.Element;
}

function MyTabPanel( { children, tabs, ...otherProps }: MyTabPanelProps ) {
	return (
		<TabPanel tabs={ tabs as any } { ...otherProps }>
			{ children as any }
		</TabPanel>
	);
}
```

And then instead of using `<TabPanel>` we use `<MyTabPanel>` (or whatever you'd rather call the component 😉)
Usually the hack doesn't take this many lines of code, but since we needed to change one of the component's sub-types it took a bit more.

Also noticed I used `ReactChild` instead of `ReactElement` for the title. That's because `ReactChild` will accept a string too.

sejas committed:
Sites Dashboard: add hover effect tab button

sejas committed:
Sites Dashboard: add hover effect tab button

wojtekn commented on packages/components/src/badge/index.tsx:
```diff
@@ -0,0 +1,23 @@
+	children: ReactNode;
+}
+
+function Badge( { children }: BadgePrpos ) {
```
We are writing a component that has the same name as the existing [Badge Calypso UI component ](https://wpcalypso.wordpress.com/devdocs/design/badge). Will it be safe?

sejas replied:
@wojtekn Thanks for finding the Badge. Yesterday I couldn't find that component.
I'll rename it and probably move it to `client/components`.

wojtekn commented:
@sejas, I found an issue (or it may be it is still in progress) related to mobile view for tabs. On the medium mobile, I can't see the "Coming soon" tab, and I can't scroll through tabs. On the large mobile view, I see "Coming soon" partially, but it is broken into two lines:

<img width="474" alt="Screen Shot 2022-07-05 at 11 23 53" src="https://user-images.githubusercontent.com/727413/177297463-066893b3-713d-4e6e-a631-81549e3088df.png">

valterlorran committed:
Design Picker: The tooltip was not displaying appropriately (#65099)

* Adding the popover as a tooltip

* Hides the outline from the tooltip

wongasy committed:
Update email upsell button to link directly to email addition page (#65058)

* Update email inbox upsell button to link directly to email addition page.

* Decoupled NewMailboxUpsell from MailboxSelectionList, generate email upsell URL by email subscription for better UX

* Simplified email subscription check using helper functions and made tweaks to improve code readability

* Adjust back/cancel button links based on context of email addition user journey to improve UX

* Adjusted upsell button url logic to improve readability

* Renamed uNewMailboxUpsell component .jsx file to .tsx

* Bind property tdomains to proper data type

* Addressed revewer feedbacks

* Use defined variable for better readability

Co-authored-by: wongasy <angela.wong@corusent.com>

danielbachhuber committed:
Me: Improve UI when photo and Gravatar profile are hidden (#65122)

* Clarify that the toggle hides both the photo and profile
* Visually indicate when photo and Gravatar profile is hidden
* Allow `border-radius: 50%;` to prevent the linter from erroring
* Clean up JS linting issues
* Use the placeholder mixin to add a loading state to `EditGravatar`
* Use the old Gravatar string until the new one has a locale translation
* Add loading UX to the Gravatar toggle element
* Tweak language the prioritize 'photo'
* Use the more precise `isFetchingUserSettings`
* Use a `min-width` to avoid slight jiggle when toggling settings
* `userSettings` constant no longer necessary after b4930f3098482171ec941c11347d593a8c90ebde
* Abstract early render paths to dedicated functions

rcanepa committed:
Jetpack Agency Dashboard: fix icon vertical misalignment (#65202)

* Use flex to vertically align icons

* Apply fix on mobile version

rcoll committed:
Jetpack Licensing Portal - Exclude certain products from the license issue screen (#64498)

* Jetpack Licensing Portal - Exclude certain products from the license issue screen

* Exclude old security plans

* Update exclude unit test

Co-authored-by: Rich Collier <rcollier@zion.local>
Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

Copons committed:
Marketplace: Add Support for Preinstalled Premium Plugins (#65125)

Add support for preinstalled premium plugins (namely: Jetpack Search) to the Dotcom Marketplace.

obenland committed:
Plugins: Add eligibility exception for Starter plans (#65215)

mmtr committed:
Revert "Plugins: Remove upgrade banners (#65123)" (#65236)

This reverts commit ab9adef7fc9c12d69be22e738ab55e6bc9d9cc60.

It brings back the upgrade banners to the Plugins page, since the lack of them could contribute to a substantial conversion loss.

ebinnion committed:
Onboarding: Fix issue with redirecting to Woo page incorrectly; Reset entire onboard store on exit (#65205)

ebinnion committed:
Revert "Onboarding: Fix issue with redirecting to Woo page incorrectly; Reset entire onboard store on exit (#65205)" (#65246)

This reverts commit 86b17bfb4a76e5bf7c4c384d9bd9b9c48cf159d9.

atanas-dev committed:
Licensing Portal: Move the required payment method step to after selecting your license product for agencies (#65229)

Agencies are required to add a payment method before they issue a license but this disrupts the ideal flow of Issue License -> Add Payment Method -> Assign License.

This change allows agencies to select a license product before adding a payment method via a short-term solution.

Context: 1202074155672006-as-1202548879851664
Future work is going to be tracked in 1201801459945917-as-1202552164778579

jeyip committed:
Editing Toolkit: Add feature flags to the help center (#65142)

valterlorran committed:
Open the dropdown links on a new tab (#65211)

p-jackson committed:
Reset to default color when user clicks "Clear" (#65107)

Co-authored-by: Daniel Bachhuber <daniel.bachhuber@automattic.com>

danielbachhuber committed:
Sites Dashboard: First pass at Sites Table design (#65251)

Co-authored-by: Daniel Bachhuber <daniel.bachhuber@automattic.com>
Co-authored-by: Antonio Sejas <antonio@sejas.es>

sirbrillig committed:
Checkout: Allow string versions of receiptId etc in getThankYouPageUrl (#65271)

* Allow string versions of receiptId etc in getThankYouPageUrl

* Add test to prevent regression

arthur791004 committed:
Fix incorrect design_type in calypso_signup_design_type_submit event (#65227)

thedebian committed:
Enable Tus uploads only for VideoPress (#64604)

* Enable Tus uploads only for videopress

* Made fileCallback more readable

* Refactor

mmtr committed:
Purchases: Show new pre-cancellation form when cancelling non-plan products (#65191)

Displays the new pre-cancellation form introduced by #56206 when cancelling other products than plans, so we can show the Atomic revert warning before cancelling a marketplace plugin from an Atomic site with a Starter plan.

robertsreberski committed:
Implement GA4 analytics for Jetpack Green (#64993)

* Add gtag ts types

* Remove redundant initializers for gtag

* Assign a valid return type to cost-to-usd

* Do not change properties in the cart project

* Flag the function with the comment about wrong type of fn property

* Implement GA4

* Implement typed version of existing function

* Remove redundant assertion (is already handled in record-purchase.js)

* Handle the param to include only specific page views in Jetpack GA

* Use USD jetpack cost in GaPurchase

* Export functions individualy rather than export object

* Allow send_page_view to be overwritten

* Add `null` to return type of costToUsd()

taipeicoder committed:
Added Hotjar and FullStory in Stepper (#65220)

phcp committed:
Goals Capture: Enable Goals Step i2 layout based on the feature flag (#65158)

* Create GoalsCaptureContainer component

It conditionally renders either the one column or two columns layout depending on the signup/goals-step-2 flag (displayAllGoals param).

* Update useGoals to return all goals if flag is on

Return all goals if the displayAllGoals is true, meaning that the goals-step-2 is active.

* Update SelectGoals component to support i2

Render bottom links if i2 is off and render the goals hint if i2 is on.

* Apply GoalsCaptureContainer on GoalsStep component

* Add selectors to support two columns layout

* Move displayAllGoals out of the GoalsStep component

escapemanuele committed:
Help Center: Improved percentage and simple sites management (#65280)

* Improved percentage and simple sites management

* Remove HC in checkout in dev

olaseni committed:
Removes the [emails/show-new-comparison-component] feature flag (#64953)

* Removes the [emails/show-new-comparison-component] feature flag

* Changed JSX to TSX

* Update JSX implementation to TSX

* Switch to more meaningful props names

* Add upsell props to define skip behaviour

* Add keys for components in a list

* Add navigation

* Fix static card data for email upsell

* Infuse upsell based design

* Fix product selection per interval by introducing new selector

* Refine upsell header text

* Modify styles

* Rename selector

* Rename argument and remove default value.

* Sort parameter order

* Mimic styles across screens

* Use string literals for keys

* Fix quantity figure

bogiii committed:
Update sites and analyzer data stores (#65232)

* Extend site data store with getting sites global styles

* Export all types from the analyzer data store

* Disable light importer flow

sejas committed:
Merge branch 'add/sites-dashboard-tabs' of github.com:Automattic/wp-calypso into add/sites-dashboard-tabs

sejas commented:
I've made a rebase of trunk and incorporated more changes than I wanted.
Closing this PR to start clean.

sejas commented:
New Clean PR: https://github.com/Automattic/wp-calypso/pull/65296
